src/colorizer/AnnotationData.ts

```diff
@@ -0,0 +1,222 @@
+import { Color } from "three";
+
+import { DEFAULT_CATEGORICAL_PALETTE_KEY, KNOWN_CATEGORICAL_PALETTES } from "./colors/categorical_palettes";
+
+import Dataset from "./Dataset";
+
+export type LabelData = {
+  name: string;
+  color: Color;
+  ids: Set<number>;
+};
+
+export interface IAnnotationData {
+  /**
+   * Returns the array of label data objects.
+   * @returns an array of `LabelData` objects, containing:
+   * - `name`: The name of the label.
+   * - `color`: The color of the label.
+   * - `ids`: A set of object IDs that have the label applied.
+   * Object properties SHOULD NOT be modified directly.
+   */
+  getLabels(): LabelData[];
+
+  /** Creates a new label and returns its index in the array of labels (as
+   * returned by `getLabels()`).
+   * @param name The name of the label. If no name is provided, a default name
+   * ("Label {number}") will be used.
+   * @param color The color to use for the label. If no color is provided, a
+   * color will be chosen from the default categorical palette.
+   * */
+  createNewLabel(name?: string, color?: Color): number;
+
+  /**
+   * Returns the indices of all labels that have been applied to the object id.
+   * (Indices are ordered as returned by `getLabels()`.)
+   * @param id The object ID to look up label indices for.
+   * @returns an array of label indices. Empty if no labels have been applied to
+   * the ID.
+   */
+  getLabelsAppliedToId(id: number): number[];
+
+  /**
+   * Returns all object IDs that the label has been applied to.
+   * @param labelIdx The index of the label to look up object IDs for.
+   * @returns an array of object IDs. Empty if the label has not been applied to
+   * any object IDs.
+   */
+  getLabeledIds(labelIdx: number): number[];
+
+  /**
+   * Returns a map from a frame number to the labeled object IDs present at that
+   * frame, represented as a record mapping label indices to IDs.
+   *
+   * The record's keys are an index of a label (as returned by `getLabels()`),
+   * and the values are arrays of object IDs present for that time frame that
+   * have that label applied.
```

Comment:
Let me know if you can think of a tidier way to word this!

Reply:
It's a hard thing to say tidily, wouldn't have commented if you didn't ask but I think this might be slightly clearer?? Just two cents, take it or leave it:

```
Returns a timeToLabelIdMap.
Each time (by frame number) maps to a record of labeled object 
IDs present at that time. The record's keys are the labels (by index), 
and the values are whatever objects (by ID) have that label applied 
at that time, as an array. 

A dataset with single frame, and a single label, and a single object 
that DOES have that label would return:
{0: 0: [0]} for {time: label: [objectWithLabel]}
```

Reply:
Oh, I like this a lot! Thank you!